### ISSUE TITLE
HPP Scanner - Multiple Issues

### ISSUE THREAD

**Issue by kingthorin** [zaproxy/zaproxy#2224]
- Contained TODO suggests switching to AbstractAppParamPlugin vs internal handling.
- Is i18n, getReference uses "extrainfo"? Should be changed to refs for clarity/consistency.
- getRisk returns 0 while bingo sets RISK_MEDIUM.

**Comment by kingthorin** [comment]
Going to tackle this in two PRs. First items 2 & 3 (plus any other items I notice along the way). Second item 1.

**Comment by kingthorin** [comment]
Ok so items 2, and 3 were simple to address.

Can someone confirm my understanding of how the existing scanner functions? (And my fwd plan.)

It seems to be that it take a shotgun approach appending the payload to all params. Take the original response, check it for INPUT fields, append the payload to all of those params, set that parameter set as part of the URL (query) for the "attack" request, submitting the request, then checking the response for the payload. Right?

So if we switch to AbstractAppParamPlugin I'm going to append the payload to the one parameter in question, ensure that parameter is on the URL (query), submit the "attack" request, check the response for the payload.

**Comment by kingthorin** [comment]
Anyone?

**Comment by psiinon** [comment]
Looks like Guifre implemented this originally, so no point in asking him now ;)
I'm not aware of any test cases for this rule. Until we have some refactoring could be tricky. 
Maybe we should demote it until we can test it properly?

**Comment by thc202** [comment]
@kingthorin Yeah, it seems to be doing that (but it also takes all of the query and POST parameters from the request, not just the INPUT fields from the response).
When using AbstractAppParamPlugin, I don't think it requires the URL check, it can be in the request body too.
For reference:
https://www.owasp.org/index.php/Testing_for_HTTP_Parameter_pollution_%28OTG-INPVAL-004%29

@psiinon I'm OK with the demotion.

**Comment by kingthorin** [comment]
Zapbot is still encountering exceptions from HPP https://github.com/zapbot/zap-mgmt-scripts/blob/66fc789cf52d78d9325d16a85b30da04fdc07785/reports/wavsep-1.5-weekly-RB-M-M.logs.txt#L637

**Comment by kingthorin** [comment]
Looks like Mutillidae might be a good testing target:
https://www.youtube.com/watch?v=Tosp-JyWVS4
https://sourceforge.net/projects/mutillidae/

**Comment by kingthorin** [comment]
The existing implementation does not seem to find the issue in Mutillidae.

**Comment by kingthorin** [comment]
I started to work on replacement scan rules, however, it quickly became apparent that something like a [Backslash Powered Scanner](https://github.com/zaproxy/zap-extensions/pull/1014) is necessary for the comparison of the test messages vs. the original. The Mutillidae test case is a poll applications so the response changes (minimally) on every submission. Which isn't very friendly for the Hirschberg approach (which has performance impacts). 

I'm going to keep this assigned to me as I do want to finish it, however it's kind of held for some other pieces. I'm also going to remove the `BasedOnZAPBOTResults` label as it's no longer really relevant.

https://groups.google.com/forum/#!searchin/zaproxy-users/pollution%7Csort:date/zaproxy-users/b5ptQZ0qFDU/AZX9wgBzBQAJ never got any responses. I think I'd like to suggest simply retiring the existing test. Thoughts on that front?
